String buffer

hwhelchel opened the pull request:
StringBuffers are used in Java to limit the number of string objects created in memory when concatenating strings.

In Ruby we can use the String#concat or String#<< methods to update strings in place so there is not a need for StringBuffers as far as I can see. It also looks like the StringIO class might serve a similar purpose.

This will add a Ruby implementation of a Java StringBuffer to emulate the behavior.

Commit:
Add string buffer data structure.

String buffer reduces runtime of concatenating strings.
Creates buffer array to store strings while concatenating.
Once all strings to be joined have been added to the buffer
the program concatenates the characters together once.

Alternatively the program might concatenate the strings for
each word in the array of strings to be joined.

Monkey patched String class to add class method called join_words.

.join_words creates a StringBuffer object and appends all words to
the buffer. Once all words have been appended it call #to_s on
the buffer which concatenates the strings.

Commit:
Revise code to update string rather than create new
string objects.

In Ruby strings are mutable. The #concat and #<<
methods allow for updating a string in place.

Alternatively the #+ method creates a new object
in memory rather than destructively altering the
string receiving the message.

edshadi commented on spec/data-structures/string-buffer/string_buffer_spec.rb:
```diff
@@ -0,0 +1,44 @@
+describe StringBuffer do
+
+  let(:my_buffer) { StringBuffer.new }
+  let(:my_word) { 'test' }
```
generally avoid naming things my_*, I think it's better if you named them buffer and word

edshadi commented on spec/data-structures/string-buffer/string_buffer_spec.rb:
```diff
@@ -0,0 +1,44 @@
+  let(:my_word) { 'test' }
+  describe '#initialize' do
+    it 'makes a StringBuffer' do
+      expect(my_buffer).to be_an_instance_of StringBuffer
```
This is testing ruby, we trust that when I initialize a class the result is then an instance of that class. I would remove that test.

edshadi commented on spec/data-structures/string-buffer/string_buffer_spec.rb:
```diff
@@ -0,0 +1,44 @@
+    context 'with a populated buffer' do
+      before(:each) { my_buffer.append my_word }
+      it 'returns a string' do
+        expect(my_buffer.to_s).to be_an_instance_of String
```
to_s always defaults to a string, you can't even force it to return anything else, try it.

edshadi commented:
looks great, I added a couple of minor comments. Let me know what you think.

Commit:
Removes EmptyBufferError. On StringBuffer creation #to_s is called.
When a StringBuffer is created with an empty buffer the program
therefore explodes. Changed it to return empty string.

Removes my_* syntax from tests.
Removes testing of instance type.

hwhelchel commented:
Hey @edshadi thanks for the comments. I updated my variable names in my tests and removed the tests which tested the data type.

In regards to #to_s, I overrode the default #to_s implementation to return the buffer array of strings as a single concatenated string. I left in this implementation but removed the testing of its data type.

edshadi commented:
nice dude!
